pyfolio/perf_attrib.py

```diff
@@ -0,0 +1,129 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import division
+import pandas as pd
+
+import empyrical
+from pyfolio.pos import get_percent_alloc
+from pyfolio.utils import print_table
+
+
+def perf_attrib(returns, positions, factor_returns, factor_loadings):
+    """
+    Does performance attribution given risk info.
+
+    Parameters
+    ----------
+    returns : pd.Series
+        Returns for each day in the date range.
+        - Example:
+            2017-01-01   -0.017098
+            2017-01-02    0.002683
+            2017-01-03   -0.008669
+
+    positions: pd.DataFrame
+        Daily holdings (in dollars or percentages), indexed by date.
+        Will be converted to percentages if positions are in dollars.
+        - Example:
+                            AAPL       TLT       XOM
+            2017-01-01  0.408046  0.534483  0.057471
+            2017-01-02  0.449367  0.101266  0.449367
+            2017-01-03  0.302632  0.282895  0.414474
+
+
+    factor_returns : pd.DataFrame
+        Returns by factor, with date as index and factors as columns
+        - Example:
+                        momentum  reversal
+            2017-01-01  0.002779 -0.005453
+            2017-01-02  0.001096  0.010290
+
+    factor_loadings : pd.DataFrame
+        Factor loadings for all days in the date range, with date and ticker as
+        index, and factors as columns.
+        - Example:
+                               momentum  reversal
+            dt         ticker
+            2017-01-01 AAPL   -1.592914  0.852830
+                       TLT     0.184864  0.895534
+                       XOM     0.993160  1.149353
+            2017-01-02 AAPL   -0.140009 -0.524952
+                       TLT    -1.066978  0.185435
+                       XOM    -1.798401  0.761549
+
+    Returns
+    -------
+    perf_attribution : pd.DataFrame
+        df with factors, common returns, and specific returns as columns,
+        and datetimes as index
+        - Example:
+                        momentum  reversal  common_returns  specific_returns
+            dt
+            2017-01-01  0.249087  0.935925        1.185012          1.185012
+            2017-01-02 -0.003194 -0.400786       -0.403980         -0.403980
+    """
+    # convert holdings to percentages, and convert positions to long format
+    positions = get_percent_alloc(positions)
+    positions = positions.stack()
+    positions.index = positions.index.set_names(['dt', 'ticker'])
+    risk_exposures = factor_loadings.multiply(positions,
+                                              axis='rows')
+    risk_exposures_portfolio = risk_exposures.groupby(level='dt').sum()
```

Comment:
We should either set the index names of `risk_exposures` ourselves or test that they exist and mark it in the doc string.